Updates for Active Learning (P1) (#326)
cisco/mindmeld: source/userguide/active_learning.rst

Active Learning in MindMeld
===========================

> Building new conversational applications, adding new domains and intents for existing applications or improving performance based on user experience, all of these benefit from the addition of more informative data to the trained models. Data generation resources and especially user logs for the application are highly beneficial in capturing new trends and growing diversity in terms of the phrasing of queries. While adding both old and new data to the trained models can improve system performance in terms of accuracy, it reduces scalability by increasing the computation and time costs.

Reviewer: I think this paragraph is a little confusing, since it sounds like we're offering a way of doing data augmentation. 
Consider something more along these lines - 

Once we have a conversational application deployed, capturing user logs are highlight beneficial to understand new trends and diversity in terms of how users are phrasing their queries for different intents. We can manually analyze these logs and annotate them with the correct domain, intent and entity labels that can then be used to train and further improve our models. However, annotation is costly and time consuming. Moreover, as the number of users for the system increase, the number of logs to be annotated also increases by many folds making manually annotation a less scalable process. In such scenarios, we can use Active Learning.

> Active Learning is an approach to continuously and iteratively learn the best data for training the models while preserving good accuracies and maintaining best-case tradeoffs with respect to computational power and time. MindMeld provides an inbuilt functionality to select the must-have queries from existing data or additional logs and datasets to get the best out of your conversational applications.

Reviewer: ```suggestion
Active Learning is an approach to continuously and iteratively pick data that is most informative for the models to train on from a pool of unlabelled data-points. By picking such data-points, active learning provides a higher chance of improving the accuracy of the model with fewer training examples. It also greatly reduces the number of queries to be annotated by humans. MindMeld provides this inbuilt functionality to select these must-have queries from existing data or additional logs and datasets to get the best out of your conversational applications.
```

Reviewer: "while preserving good accuracies and maintaining best-case tradeoffs with respect to computational power and time" - don't think this is something we can guarantee.

Author: Makes sense, higher chance sounds reasonable!

| train_pattern          | str    | Regex pattern to match train files. Example for all train files ".*train.*.txt"  |
+------------------------+--------+----------------------------------------------------------------------------------+
| test_pattern           | str    | Regex pattern to match train files. Example for all test files ".*test.*.txt"    |
+------------------------+--------+----------------------------------------------------------------------------------+
| train_seed_pct         | float  | Percentage of training data to use as the initial seed                           |
+------------------------+--------+----------------------------------------------------------------------------------+

3. **tuning**

+------------------------+------------+----------------------------------------------------------------------------------+
| Configuration Key      | Type       | Definition and Examples                                                          |
+========================+============+==================================================================================+

> | n_classifiers          | int        | Number of classifier instances to be used by a subset of strategies              |

Reviewer: Number of classifier instances to be used by ensemble strategies?

+------------------------+------------+----------------------------------------------------------------------------------+
| n_epochs               | int        | Number of turns strategy tuning is run on the complete log data                  |
+------------------------+------------+----------------------------------------------------------------------------------+
| batch_size             | int        | Number of queries to sample at each iteration from the logs                      |
+------------------------+------------+----------------------------------------------------------------------------------+

> | tuning_level           | str        | Selecting classifiers to use for tuning, options: "domain", "intent" or "joint"  |

Reviewer 2: We can remove "joint" here as we only support "domain" and "intent"

+------------------------+------------+----------------------------------------------------------------------------------+
| labeled_logs_pattern   | str        | Regex pattern to match log files if already labeled. For example, ".*log.*.txt"  |
+------------------------+------------+----------------------------------------------------------------------------------+
| unlabeled_logs_path    | str        | Path to text file containing unlabeled queries from user logs or other resources |
+------------------------+------------+----------------------------------------------------------------------------------+

If there is no configuration defined in the ``config.py`` file or if fields are missing in the custom configuration, the relevant missing information is obtained from the following default configuration:

.. code-block:: python

    DEFAULT_ACTIVE_LEARNING_CONFIG = {
        "output_folder": None,
        "pre_tuning": {
            "train_pattern": ".*train.*.txt",
            "test_pattern": ".*test.*.txt",
            "train_seed_pct": 0.20,
        },
        "tuning": {
            "n_classifiers": 3,
            "n_epochs": 5,

>             "batch_size": 100,

Reviewer 2: Would be good to add a warning or note somewhere that the default batch size is 100. For large applications, this number may be too small and we should encourage developers to update it accordingly. (Recommend 1-2% of total training data size, which is what most papers seem to do)

Strategies
^^^^^^^^^^
The tuning step allows the application to run 7 possible strategies and choose the best performing one. Each strategy is a sampling function that samples the worst performing queries from the latest batch iteration of the training data. The assessment of worst performance comes from the classifiers' confidence in the predictions for that query. All heuristics use this information differently as described next.

+---------------------------+-----------------------------------------------------------------------------------------------+
| Strategy                  | How does it work?                                                                             |
+===========================+===============================================================================================+
| Random Sampling           | Samples the next set of queries at random                                                     |
+---------------------------+-----------------------------------------------------------------------------------------------+
| Least Confidence Sampling | From the available queries in the batch, this sampling strategy samples queries with the      |

> |                           | lowest max confidence score across any class, i.e., queries that are the least confident about|

Reviewer: ```suggestion
|                           | lowest max confidence score across any class, i.e., queries that the classifier is least confident about|
```

|                           | any of the available classes in the classifier.                                               |
+---------------------------+-----------------------------------------------------------------------------------------------+
| Entropy Sampling          | Calculates the entropy score of the classifier confidences per query. Samples the ones with   |
|                           | highest entropy.                                                                              |
+---------------------------+-----------------------------------------------------------------------------------------------+

> | Margin Sampling           | From the available queries in the batch, this sampling strategy samples queries with the      |

Reviewer: ```suggestion
| Margin Sampling           | From the available queries in the batch, this sampling strategy samples queries      |
```

|                           | every query (% of classifiers that voted for the most frequent class). The queries are then   |
|                           | ranked from lowest classifier agreement to highest and the sampled in order.                  |
+---------------------------+-----------------------------------------------------------------------------------------------+
| KL Divergence Sampling    | Across n runs of the classifier, this sampling strategy calculates the KL divergence between  |
|                           | average confidence distribution across all classifiers for a given class and the confidence   |
|                           | distribution for a given query for that class. Queries with higher divergence are sampled.    |
+---------------------------+-----------------------------------------------------------------------------------------------+
| Ensemble Sampling         | Combines ranks from all the above heuristics and samples in order.                            |
+---------------------------+-----------------------------------------------------------------------------------------------+

Tuning Levels
^^^^^^^^^^^^^

> Tuning level defines the level at which the classifier evaluation is done and for which the confidence scores or model probabilites are passed onto the strategies defined above. One of three strategies can be employed - domain, intent or joint.

Reviewer: ```suggestion
Since MindMeld defines a hierarchy of domains and intents, the various heuristics can be computed by using the confidence scores or probabilities of either the domain or intent classifiers or both. This level  is indicated by the Tuning level in the config. 
```

> 1. For the domain level, the domain classifier is run and the probability scores for the domain classifier are passed to the strategies.

Reviewer: ```suggestion
1. For the domain level, the domain classifier is run and the probability scores of the classifier are passed to the strategies.
```

> 2. For the intent level, the intent classifier scores (vectors) across all domains are compressed down to a single vector and passed on to the strategies.

Reviewer: ```suggestion
2. For the intent level, the intent classifier probability scores across all domains are concatenated into a single vector and passed on to the strategies.
```

Reviewer: Alternatively, can also use `combined` instead of `compressed`.

> 3. For the joint level, a dot product is taken across the domain and intent level predictions and the combined representation of confidences/probability scores is provided to the strategies.

Reviewer: ```suggestion
3. For the joint level, a dot product is taken across the domain and all intent level predictions and the combined representation of confidences/probability scores is provided to the strategies.
```

> 3. For the joint level, a dot product is taken across the domain and intent level predictions and the combined representation of confidences/probability scores is provided to the strategies.

Reviewer: ```suggestion
3. For the joint level, a dot product is taken across all the domain and intent level predictions and the combined representation of confidences/probability scores is provided to the strategies.
```

Strategy Tuning
---------------

> The strategy tuning phase in the active learning pipeline is useful in choosing the best tuning strategies for your application. According to the config value of ``train_seed_pct``, the pipeline picks that percentage of training data (identified as filed that match the ``train_pattern`` regex in the config file) as seed data. Next, the classifiers are trained on this seed data and evaluated on the test data (files matching ``test_pattern``). Iteratively, data equivalent to the ``batch_size`` is added from the remaining training data to the initial seed data and the classifier models retrained. This selection is made according to different tuning strategies which are discussed later. The test or evaluation dataset remains static. In this process the train set keeps growing till all training data has been consumed by the model and the final iteration of classifier training has been trained on all possible training data. The accuracy, query selection and tuning strategy performance results are saved along with performance plots tracking the classifier performance in tandem with the increasing training data. This process is repeated for ``n_epochs`` (as defined in the config) to obtain average active learning performance. Results obtained at the end can be used to quantitatively and visually choose the best tuning approach for your application.

Reviewer: ```suggestion
The strategy tuning phase in the active learning pipeline is useful in choosing the best tuning strategies for your application. We partition the existing training data in the app as a seed set and unsampled set according to the  ``train_seed_pct`` value mentioned in the config file. Note that the pipeline only uses the data from files that match the ``train_pattern`` regex in the config file in this step.  Next, the classifiers are trained on this seed data and evaluated on the existing test data (files matching ``test_pattern``). Iteratively, data equivalent to the ``batch_size`` is added from the remaining unsampled training data to the initial seed data by using the different sampling strategies discussed earlier. The classifier models are retrained with the new expanded seed set and evaluated against the test set. In this process the train set keeps growing till all training data has been consumed by the model and the final iteration of classifier training has been trained on all possible training data. 

The accuracy, query selection and tuning strategy performance results are saved along with performance plots tracking the classifier performance in tandem with the increasing training data. We repeat the above Tuning process for ``n_epochs`` (as defined in the config) to obtain average active learning performance. Results obtained at the end can be used to quantitatively and visually choose the best tuning approach for your application.
```

The following command can be used to run query selection using the settings defined in the application configuration if the log file or the log files' pattern has been specified in the config:

.. code-block:: console

    mindmeld active_learning --select --app-path '<PATH>/app_name/' --output_folder '<PATH>'

Alternatively, path to unlabeled logs (``unlabeled_logs_path``) can be provided as a flag. 

.. code-block:: console

    mindmeld active_learning --select --app-path "<PATH>/app_name/" --output_folder '<PATH>' --unlabeled_logs_path "<PATH>/logs.txt"

Also, if your log data is labelled and included in your MindMeld application you can specify the pattern for your log data using the following flag:

.. code-block:: console 

    mindmeld active_learning --select --app-path '<PATH>/app_name/' --output_folder '<PATH>' --labeled_logs_pattern ".*log.*.txt"

Optional flags that can be used for selection include: ``batch_size``, ``log_usage_pct``, ``strategy``.

.. note::

>     For labelled logs pattern, ensure that your log pattern (``labeled_logs_pattern``) does not overlap with your train pattern (``train_pattern``).

Reviewer 2: Would be helpful to have a quick example here of what we mean by "overlapping pattern"

>     For labelled logs pattern, ensure that your log pattern (``labeled_logs_pattern``) does not overlap with your train pattern (``train_pattern``).

Reviewer: ```suggestion
    When selecting from labelled logs, ensure that your log pattern (``labeled_logs_pattern``) does not overlap with your train pattern (``train_pattern``).
```

Active Learning in MindMeld
===========================

> Once a conversational application is deployed, capturing user logs can be highly beneficial to understand new trends and diversity in terms of how users are phrasing their queries for different intents. These logs can improve the quality of the training data and overall classifier performance. However, manually analyzing these logs and annotating them with the correct domain, intent and entity labels is costly and time consuming. Moreover, as the number of users for the system increase, the number of logs to be annotated also increases by many folds. This reduces the scalability of the manual annotation process. For such scenarios, we can use Active Learning.

Reviewer 3: Once a conversational application is deployed, capturing user logs can be highly beneficial to understand new trends and diversity of how users are phrasing their queries for different intents. These logs can improve the quality of the training data and overall classifier performance. However, manually analyzing these logs and annotating them with the correct domain, intent and entity labels is costly and time consuming. Moreover, as the number of users for the system increase, the number of logs to be annotated also increases by many folds. This reduces the scalability of the manual annotation process. For such scenarios, we can use Active Learning.

Reviewer: ```suggestion
Once a conversational application is deployed, capturing user logs can be highly beneficial to understand new trends and diversity of how users are phrasing their queries for different intents. These logs can improve the quality of the training data and overall classifier performance. However, manually analyzing these logs and annotating them with the correct domain, intent and entity labels is costly and time consuming. Moreover, as the number of users of the system increase, the number of logs to be annotated also increases by many folds. This reduces the scalability of the manual annotation process. In such scenarios, we can use Active Learning.
```

What is Active Learning?
^^^^^^^^^^^^^^^^^^^^^^^^
In this section, we take a closer look at what active learning is and how it works within MindMeld.

Active Learning is an approach to continuously and iteratively pick data that is most informative for the models to train on from a pool of data-points. By choosing these informative data-points, active learning provides a higher chance of improving the accuracy of the model with fewer training examples. If the pool of data-points is unannotated, active learning also greatly reduces the number of queries that need manual annotation. MindMeld provides this inbuilt functionality to select these must-have queries from existing data or additional logs and datasets to get the best out of your conversational applications.

> MindMeld's NLP pipeline consists of a heirarchy of classifier and tagger models for domain, intent, entity and role classification. Each classifier takes as input the query text and extracted features to generate a probability distribution for the corresponding classes in the heirarchy. These probability distributions can be strong indicator of the classifier's confidence in its predictions and can be useful in determining whether the query needs further reflection. 

Reviewer: ```suggestion
MindMeld's NLP pipeline consists of a hierarchy of classifier and tagger models for domain, intent, entity and role classification. Each classifier takes as input the query text and extracted features to generate a probability distribution for the corresponding classes in the hierarchy. These probability distributions can be strong indicators of the classifier's confidence in its prediction and can be useful in determining whether the query needs further reflection. 
```

> Say for a query A from the logs, the selected class (or the highest probability class) in the domain classifier is at 0.95 or 95% and the remaining 5% is distributed amongst the rest of the domains, this could potentially indicate a high confidence by the classifier in its prediction. Whereas a query B has a probability 0.5 for the selected class, possibly indicating lower prediction confidence. Based on these values, the active learning strategies will select query B over query A for adding back to the training data. With query B now in the training data with proper annotation, the system has a higher chance of performing better on similar confusable queries. Later sections explain how different strategies use these confidence scores to select important queries.

Reviewer: ```suggestion
Say for a query A from the logs, the classifier assigns a probability of 0.95 or 95% to the selected class (i.e, the highest probability class) in the domain classifier and the remaining 5% is distributed amongst the rest of the domains. This indicates a high confidence by the classifier in its prediction. Whereas, for a query B, if the classifier assigns a probability 0.5 for the selected class, it would indicate a lower confidence in its prediction. Based on these confidence values, certain active learning strategies will select query B over query A which can be annotated and added to the training data. Once the classifier is trained with this additional data, it has a higher chance of performing better on similar confusable queries. Later sections explain how different strategies use these confidence scores to select important queries.
```

> There are two phases in MindMeld's active learning pipeline. One for configuring the best hyperparamters (such as optimal strategy) for the pipeline and the other for selecting the most important subset of queries from logs given the configured pipeline. These phases are referred to as ``Strategy Tuning`` and ``Query Selection``. Before diving deeper into these phases, let us take a look at the configuration for setting up MindMeld's active learning pipeline.

Reviewer: ```suggestion
There are two phases in MindMeld's active learning pipeline. One, for configuring the best hyperparameters (such as the optimal strategy) for the pipeline and the other, for selecting the most important subset of queries from logs given the optimal configuration. These phases are referred to as ``Strategy Tuning`` and ``Query Selection`` respectively. Before diving deeper into these phases, let us take a look at the configuration file for setting up MindMeld's active learning pipeline.
```

.. _al_config:

Defining the AL config
^^^^^^^^^^^^^^^^^^^^^^

> The first step towards customizing the active learning pipeline to your application is configuration adjustment. This section details different components in the configuration and provides a default configuration setup which is applied if no active learning config is defined in ``config.py``. There are five components in this configuration:

Reviewer: ```suggestion
Modifying the configuration file is the first step towards customizing the active learning pipeline to your application. This section details different components in the configuration along with the default values that are applied if no `ACTIVE_LEARNING_CONFIG` (?) is defined in ``config.py``. There are five main components:
```

Reviewer: Not sure what the key we're expecting is. (`ACTIVE_LEARNING_CONFIG` (?) )

1. **output_folder** (str) - Directory path to save the output of the active learning pipeline.

2. **pre_tuning**

+---------------------------+------------+----------------------------------------------------------------------------------+
| Configuration Key         | Type       | Definition and Examples                                                          |
+===========================+============+==================================================================================+
| train_pattern             | str        | Regex pattern to match train files. Example for all train files ".*train.*.txt"  |
+---------------------------+------------+----------------------------------------------------------------------------------+

> | test_pattern              | str        | Regex pattern to match train files. Example for all test files ".*test.*.txt"    |

Reviewer: ```suggestion
| test_pattern              | str        | Regex pattern to match test files. Example for all test files ".*test.*.txt"    |
```

+---------------------------+------------+----------------------------------------------------------------------------------+
| train_seed_pct            | float      | Percentage of training data to use as the initial seed                           |
+---------------------------+------------+----------------------------------------------------------------------------------+

3. **tuning**

+---------------------------+------------+----------------------------------------------------------------------------------+
| Configuration Key         | Type       | Definition and Examples                                                          |
+===========================+============+==================================================================================+
| n_classifiers             | int        | Number of classifier instances to be used by ensemble strategies                 |
+---------------------------+------------+----------------------------------------------------------------------------------+

> | n_epochs                  | int        | Number of turns strategy tuning is run on the complete log data                  |

Reviewer: Shouldn't it be train data and not log data?

+---------------------------+------------+----------------------------------------------------------------------------------+

> | batch_size                | int        | Number of queries to sample at each iteration from the logs                      |

Reviewer: Same comment as above. We're not operating on (un)annotated logs during the tuning phase, right?

+---------------------------+------------+----------------------------------------------------------------------------------+

> | tuning_level              | str        | Selecting classifiers to use for tuning, options: "domain" or "intent"           |

Reviewer: ```suggestion
| tuning_level              | str        | Selecting the hierarchy level to use for tuning, options: "domain" or "intent"           |
```

+---------------------------+------------+----------------------------------------------------------------------------------+
| tuning_strategies         | List[str]  | List of strategies (heuristics) to use for tuning                                |
+---------------------------+------------+----------------------------------------------------------------------------------+

4. **tuning_output**

+---------------------------+------------+----------------------------------------------------------------------------------+
| Configuration Key         | Type       | Definition and Examples                                                          |
+===========================+============+==================================================================================+

> | save_sampled_queries      | bool       | Whether to save the queries sampled at each iteration                            |

Reviewer: ```suggestion
| save_sampled_queries      | bool       | Option to save the queries sampled at each iteration                            |
```

+---------------------------+------------+----------------------------------------------------------------------------------+

> | aggregate_statistic       | str        | Metric to record aggregate classifier performance data in the output and plots,  |

Reviewer: ```suggestion
| aggregate_statistic       | str        | Aggregated classifier performance metric to record in the output and plots,  |
```

|                           |            | options: "accuracy", "f1_weighted", "f1_macro", "f1_micro"                       |
+---------------------------+------------+----------------------------------------------------------------------------------+

> | class_level_statistic     | str        | Metric to record class/tuning level performance data in the output and plots,    |

Reviewer: ```suggestion
| class_level_statistic     | str        | Class level performance metric to record in the output and plots,    |
```

|                           |            | options: "f_beta", "percision", "recall"                                         |
+---------------------------+------------+----------------------------------------------------------------------------------+

5. **query_selection**

+---------------------------+------------+----------------------------------------------------------------------------------+
| Configuration Key         | Type       | Definition and Examples                                                          |
+===========================+============+==================================================================================+
| selection_strategy        | str        | Strategy (heuristic) to use for log selection                                    |
+---------------------------+------------+----------------------------------------------------------------------------------+
| log_usage_pct             | float      | Percentage of the log data to use for selection                                  |
+---------------------------+------------+----------------------------------------------------------------------------------+

> | labeled_logs_pattern      | str        | Regex pattern to match log files if already labeled. For example, ".*log.*.txt"  |

Reviewer: Should we mention here that annotated log files just mean that these files have been dumped appropriately in the hierarchy or is that understood?

.. code-block:: python

    DEFAULT_ACTIVE_LEARNING_CONFIG = {
        "output_folder": None,
        "pre_tuning": {
            "train_pattern": ".*train.*.txt",
            "test_pattern": ".*test.*.txt",
            "train_seed_pct": 0.20,
        },
        "tuning": {
            "n_classifiers": 3,
            "n_epochs": 5,
            "batch_size": 100,
            "tuning_level": "domain",
            "tuning_strategies": [
                "LeastConfidenceSampling",
                "MarginSampling",
                "EntropySampling",
                "RandomSampling",
                "DisagreementSampling",
                "EnsembleSampling",
                "KLDivergenceSampling",
            ],
        },
        "tuning_output": {
            "save_sampled_queries": True,
            "aggregate_statistic": "accuracy",
            "class_level_statistic": "f_beta",
        },
        "query_selection": {
            "selection_strategy": "EntropySampling",
            "log_usage_pct": 1.00,
            "labeled_logs_pattern": None,
            "unlabeled_logs_path": "logs.txt",
        },
    }

.. note::

>     The default batch size is 100. For large applications, this number may be too small and we should encourage developers to update it accordingly. We recommend setting 1-2% of total training data size as the batch size.

Reviewer: ```suggestion
    The default batch size is 100. For large applications, this number may be too small and we encourage developers to update it accordingly. We recommend setting 1-2% of total training data size as the batch size.
```

.. _al_strategy_tuning:

Strategy Tuning
^^^^^^^^^^^^^^^

> The goal of the strategy tuning phase in the active learning pipeline is to determine the best tuning strategies and tuning level for your application. We will talk about the different possible tuning strategies and levels later in this section. 

Reviewer: ```suggestion
The goal of the strategy tuning phase in the active learning pipeline is to determine the best strategy (heuristic) and tuning level for your application. We will talk about the different possible strategies and tuning levels later in this section. 
```

.. image:: /images/strategy_tuning.png
    :align: center
    :name: strategy_tuning_flow

> The existing training data in the app is partitioned into a `sampled` seed set and `unsampled` set according to the  ``train_seed_pct`` value mentioned in the config file. Data is evenly sampled at the ``train_seed_pct`` value across the domains to maintain class balance in the seed dataset. Note that the pipeline only uses the data from files that match the ``train_pattern`` regex in the config file in this step. The classifiers are trained on this sampled seed data (and evaluated on the existing test data, i.e. files matching ``test_pattern``).

Reviewer: ```suggestion
In this phase, the existing training data in the app is partitioned into a `sampled` seed set and `unsampled` set according to the  ``train_seed_pct`` value mentioned in the config file. Data is evenly sampled based on given ``train_seed_pct`` across the different domains/intents to maintain class balance in the seed dataset. Note that the pipeline only uses the data from files that match the ``train_pattern`` regex in the config file in this step. The classifiers are trained on this sampled seed data (and evaluated on the existing test data, i.e. files matching ``test_pattern``).
```

> Next, the trained classifiers are used to generate a prediction for queries in the unsampled set. This prediction is output in the form of a class probability distribution. This distribution is passed through the sampling or ``tuning_strategies`` to obtain a ranked list of unsampled queries. Based on this ranked list, the top-k queries (k = ``batch_size``) are extracted from the unsampled set and added to the sampled set. Thereby increasing the size of the latter while reducing that of the former.

Reviewer: ```suggestion
Next, the trained classifiers are used to generate a predictions for queries in the unsampled set. These predictions are output in the form of class probability distributions. The various ``tuning_strategies`` use these distributions to rank the queries. Based on this ranked list, the top-k queries (k = ``batch_size``) are extracted from the unsampled set and added to the sampled set, thereby increasing the size of the latter while reducing that of the former.
```

The classifier models are now retrained with the expanded sampled set and evaluated against the same test set. This process is repeated until all the unsampled training data has been consumed by the sampled set and the final iteration of classifier training is done on this exhaustive sampled set. This tuning process is repeated for ``n_epochs`` (as defined in the config) to obtain average active learning performance.

Information about unsampled queries selected at each iteration, and the performance of classifiers for each tuning strategy is stored iteratively in the ``output_folder`` directory. For every tuning command run, a new experiment folder is generated in this directory with the performance results and corresponding plots for a better visual understanding of the results.

The following command can be used to run tuning using the settings defined in the application configuration:

.. code-block:: console

    mindmeld active_learning --tune --app-path '<PATH>/app_name/' --output_folder '<PATH>'

> Flags for application path and output folder are required and overwrite the default configuration settings for active learning. In addition to the aforementioned required flags, the following optional flags can be used - tuning_level, batch_size, n_epochs, train_seed_pct, and plot (default ``True``). These are described in detail in AL config section above.

Reviewer: If they are required fields, they should not have defaults?

Author: Yeah the two don't? Rest are optional.

Reviewer: "Flags for application path and output folder are required and overwrite the default configuration settings for active learning." --> "Flags for application path and output folder are required. In addition to the aforementioned required flags, the following optional flags can be used - tuning_level, batch_size, n_epochs, train_seed_pct, and plot (default ``True``). These are described in detail in AL config section above. These flags overwrite the default configuration settings for active learning."

> At the end of the tuning process, results are stored in the ``output_folder``. The ``accuracy.json`` file in the directory ``output_folder/results`` consist of strategy performance on the application's test/evaluation data for every iteration and epoch. ``selected_queries.json`` consists of the same information but instead of evaluation performance, this file records the queries selected at that iteration. the ``output_folder/plots`` directory consists of this quantitative information in a  visual format. The plots record performance of all chosen strategies across data iterations and gives a sense of which strategy is best suited for your application. The same information can be gauged from these results and plots about the best ``tuning_level`` for your application.

Reviewer: ```suggestion
At the end of the tuning process, results are stored in the ``output_folder``. The ``accuracy.json`` file in the directory ``output_folder/results`` consist of strategy performance on the application's test/evaluation data for every iteration and epoch. ``selected_queries.json`` consists of the same information but instead of evaluation performance, this file records the queries selected for that iteration. The ``output_folder/plots`` directory consists of the same quantitative information in a visual format. The plots record performance of all chosen strategies across iterations and give a sense of which strategy is best suited for your application. The same information can be gauged from these results and plots about the best ``tuning_level`` for your application.
```

| Entropy Sampling          | Calculates the entropy score of the classifier confidences per query. Samples the ones with   |
|                           | highest entropy.                                                                              |
+---------------------------+-----------------------------------------------------------------------------------------------+
| Margin Sampling           | From the available queries in the batch, this sampling strategy samples queries that have the |
|                           | lowest confidence score difference between the top two class confidence scores for the query. |
|                           | This difference is referred to as the "margin".                                               |
+---------------------------+-----------------------------------------------------------------------------------------------+
| Disagreement Sampling     | Across n runs of the classifier, this sampling strategy calculates an agreement score for     |
|                           | every query (% of classifiers that voted for the most frequent class). The queries are then   |

> |                           | ranked from lowest classifier agreement to highest and the sampled in order.                  |

Reviewer: ```suggestion
|                           | ranked from lowest classifier agreement to highest and then sampled in order.                  |
```

| KL Divergence Sampling    | Across n runs of the classifier, this sampling strategy calculates the KL divergence between  |
|                           | average confidence distribution across all classifiers for a given class and the confidence   |
|                           | distribution for a given query for that class. Queries with higher divergence are sampled.    |
+---------------------------+-----------------------------------------------------------------------------------------------+
| Ensemble Sampling         | Combines ranks from all the above heuristics and samples in order.                            |
+---------------------------+-----------------------------------------------------------------------------------------------+

.. _tuning_levels:

Tuning Levels
-------------
Since MindMeld defines a hierarchy of domains and intents, the various heuristics can be computed by using the confidence scores or probabilities of either the domain or intent classifiers. This level is indicated by the ``tuning_level`` in the config. 

* For the domain level, the domain classifier is run and the probability scores of the classifier are passed to the strategies.

> * For the intent level, the intent classifier probability scores across all domains are combined into a single vector and passed on to the strategies.

Reviewer: Maybe we should say "concatenated" instead of "combined"?

> Once the tuning step has been completed and the results observed, a decision can be made on the best performing hyperparameters, strategy and tuning level, for the query selection step. We explore this step next.

Reviewer: ```suggestion
Once the tuning step has been completed and the results observed, a decision can be made on the best performing hyperparameters, strategy and tuning level for the query selection step. 
```

.. _query_selection:

Query Selection
^^^^^^^^^^^^^^^
Having obtained optimized hyperparameters through the tuning step, the pipeline is ready to run the query selection step. Here, the active learning pipeline picks the best subset of queries from the logs that can be added to the training files to give the maximum performance boost in terms of accuracy.

.. image:: /images/query_selection.png
    :align: center
    :name: query_selection_flow

Two sets of data inputs are needed for the query selection step, application (train and test) data and user logs. The user logs can either be additionally annotated log files within the MindMeld application hierarchy (see ``labeled_logs_pattern`` in config) or an external text file consisting of log queries (``unlabeled_logs_path``). After processing the data through the active learning data loader, the train data and log data are obtained. Additionally, ``log_usage_pct`` is a configuration setting that can be used to determine what amount of the user logs should be considered for selection. By default all log data is considered available for selection.

> At this point, the MindMeld NLP model is trained once using the train data. This model is then used to infer predictions on the log data and generate classifier probability distributions for all queries. Note that the classification model used at this step is the ``tuning_level`` determined to be the best performing one at the tuning step.

Reviewer: ```suggestion
At this point, the MindMeld classifier is trained using the train data. This model is then used to infer predictions on the log data and generate classifier probability distributions for all queries. Note that the classification model (domain or intent) used at this step is specified by the ``tuning_level`` determined to be the best performing one in the tuning step.
```

At this stage, the queries can be annotated and added to the train files of the MindMeld application. Next, we take a look at some options to run the query selection process.

The following command can be used to run query selection using the application's configuration, if the log file or the log files' pattern has been specified in the config:

.. code-block:: console

    mindmeld active_learning --select --app-path '<PATH>/app_name/' --output_folder '<PATH>'

Alternatively, path to unlabeled logs text file (``unlabeled_logs_path``) can be provided as a flag. 

.. code-block:: console

    mindmeld active_learning --select --app-path "<PATH>/app_name/" --output_folder '<PATH>' --unlabeled_logs_path "<PATH>/logs.txt"

Also, if your log data is labelled and included in your MindMeld application you can specify the pattern for your log data using the following flag:

.. code-block:: console 

    mindmeld active_learning --select --app-path '<PATH>/app_name/' --output_folder '<PATH>' --labeled_logs_pattern ".*log.*.txt"

Optional flags that can be used for selection include: ``batch_size``, ``log_usage_pct``, ``strategy``.

.. note::

>     When selecting from labelled logs, ensure that the regex pattern provided in log pattern (``labeled_logs_pattern``) don't have overlapping regex pattern in (``train_pattern``). In other words, ensure that the same files are not chosen by the system for both train and log data.

Reviewer: no overlap with train and test patterns

Quick Reference
^^^^^^^^^^^^^^^
This section is a quick reference on the basic command-line usage of the active learning Strategy Tuning (``tune``) and Query Selection (``select``) commands. Refer the earlier sections for additional flags.

Strategy tuning

.. code-block:: console

    mindmeld active_learning --tune --app-path '<PATH>/app_name/' --output_folder '<PATH>'

Query selection

.. code-block:: console

    mindmeld active_learning --select --app-path "<PATH>/app_name/" --output_folder '<PATH>' --unlabeled_logs_path "<PATH>/logs.txt"

.. note::
    

>     * Running these commands without defining a custom active learning configuration in ``config.py`` would result in the use of a default configuration. The custom configuration settings and MindMeld's default active learning configuration are explained in the next section.

Reviewer: ```suggestion
    * Running these commands without defining a custom active learning configuration in ``config.py`` would result in the use of a default configuration. The custom configuration settings and MindMeld's default active learning configuration are explained in the <link to section>
```
